packages/suite/src/actions/wallet/discoveryActions.ts

```diff
@@ -269,6 +279,82 @@ export const updateNetworkSettings = () => (dispatch: Dispatch, getState: GetSta
     });
 };
 
+const getAvailableCardanoDerivations =
+    (deviceState: string, device: TrezorDevice) =>
+    async (dispatch: Dispatch): Promise<('normal' | 'legacy' | 'ledger')[] | undefined> => {
+        // If icarus and icarus-trezor derivations return same pub key
+        // we can skip derivation of the latter as it would discover same accounts.
+        // Ledger derivation will always result in different pub key except in shamir where all derivations are the same
+        const commonParams = {
+            device,
+            keepSession: true,
+            useEmptyPassphrase: device.useEmptyPassphrase,
+            path: "m/1852'/1815'/0'",
+        };
+        const icarusPubKeyResult = await TrezorConnect.cardanoGetPublicKey({
+            ...commonParams,
+            derivationType: getDerivationType('normal'),
+        });
+
+        const icarusTrezorPubKeyResult = await TrezorConnect.cardanoGetPublicKey({
+            ...commonParams,
+            derivationType: getDerivationType('legacy'),
+        });
+
+        const ledgerPubKeyResult = await TrezorConnect.cardanoGetPublicKey({
+            ...commonParams,
+            derivationType: getDerivationType('ledger'),
+        });
+
+        if (
+            icarusPubKeyResult.success &&
+            icarusTrezorPubKeyResult.success &&
+            ledgerPubKeyResult.success
+        ) {
+            const icarusPubKey = icarusPubKeyResult.payload.publicKey;
+            const icarusTrezorPubKey = icarusTrezorPubKeyResult.payload.publicKey;
+            const ledgerPubKey = ledgerPubKeyResult.payload.publicKey;
+
+            if (icarusPubKey === icarusTrezorPubKey && icarusPubKey === ledgerPubKey) {
+                // all pub keys are the same
+                return ['normal'];
+            }
+            if (icarusPubKey === icarusTrezorPubKey) {
+                // ledger pub key is different
+                return ['normal', 'ledger'];
+            }
+
+            // each pub key is different
+            return ['normal', 'legacy', 'ledger'];
+        }
+
+        let error: string | undefined;
+        let code: string | undefined;
+
+        // extract error from first failed cardanoGetPublicKey request
+        const derivationFail = [
+            icarusPubKeyResult,
+            icarusTrezorPubKeyResult,
+            ledgerPubKeyResult,
+        ].find(r => !r.success);
+        if (derivationFail && !derivationFail.success) {
+            error = derivationFail.payload.error;
+            code = derivationFail.payload.code;
+        }
+
+        dispatch(
+            update(
+                {
+                    deviceState,
+                    status: DISCOVERY.STATUS.STOPPED,
+                    error,
+                    errorCode: code,
+                },
+                DISCOVERY.STOP,
+            ),
+        );
```

Comment:
I am not sure under what conditions could calling `cardanoGetPublicKey` fail. I assume it pretty much won't happen, but if it does the discovery fails with an error from first failed call and the error it will be shown to the user.

Reply:
Probably only some kind of usb communication interruption could cause this.